Tidy up the Coq package sets

vbgl opened the pull request:
###### Motivation for this change

See discussion of #30698.
cc/ @maximedenes @jwiegley @Zimmi48 

There are three (3) `coqPackages_XX` set, one for each version of Coq (8.5, 8.6, 8.7); there are all built from the same function `mkCoqPackages`.

Before this PR, this sets all have the same attributes but many libraries are not compatible with all these versions of Coq. Therefore, many derivations in these sets are broken: they either fail at evaluation or at build time. It makes difficult to know what are the available libraries in `nixpkgs` and whether something is broken by mistake or by design.

The goal of this PR is that all attributes evaluate and build. Here is how it is achieved:

 0. Move most Coq-related stuff in a dedicated file `pkgs/top-level/coq-packages.nix`, available through the `coqLib` attribute.
 1. Add to most Coq library derivations a `meta.compatibleCoqVersions` attributes: it is a function from version strings to booleans and tells whether said version is compatible with said library. If this attribute is missing, it means that the library is compatible with all versions (i.e., defaults to `_: true`).
 2. At the end of the fixed point computation, the (recursive) attribute set `coqPackages_XX` is filtered to keep only those attributes that are compatible with the corresponding Coq version.

NB: I intended to do the filtering during the fixed point computation, but run into infinite loops. It’s probably not a big deal but I wished I understood what’s going on…

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

vbgl committed:
coqPackages: move to a separate file and filter the package set

vbgl committed:
coq: minor cleaning

Zimmi48 commented:
This looks good to me. And I've done some testing under NixOS (with sandboxing).

Given that the predicate is applied to the `coq-version` attribute, I would rather prefer a definition of the form `compatibleCoqVersions = v: builtins.elem v [ "8.6" "8.7" ];` that has IMO the advantage of being explicit.

I suppose that what grahamcofborg-eval-check-meta shows is that this information should not be in the `meta` attribute set.

orivej commented on pkgs/top-level/coq-packages.nix:
```diff
@@ -0,0 +1,76 @@
+    };
+
+  filterCoqPackages = coq:
+    lib.filterAttrsRecursive
```
Is `lib.filterAttrs` insufficient?

orivej commented on pkgs/top-level/all-packages.nix:
```diff
@@ -18824,57 +18824,13 @@ with pkgs;
-  coqPackages_8_7 = mkCoqPackages coqPackages_8_7 coq_8_7;
-  coqPackages = coqPackages_8_6;
-  coq = coqPackages.coq;
+  inherit (coqLib)
```
May you `inherit (callPackage ./coq-packages.nix {})` to avoid confusing naming. (Due to `coqLib = ./coq-packages ≠ coqPackages`.)

grahamc commented on pkgs/top-level/coq-packages.nix:
```diff
@@ -0,0 +1,76 @@
+      let pred =
+        if p ? meta && p.meta ? compatibleCoqVersions
+        then p.meta.compatibleCoqVersions
+        else _: true;
```
Note that compatibleCoqVersions isn't a valid meta attribute. Maybe `passthru` would be more appropriate?

vbgl commented:
Thanks for the comments. I addressed your concerns in 791776b.

Do you know if there is a way to inherit *all* attributes from a set? Currently the line `inherit (callPackage ./coq-packages.nix {}) …;` repeats all the attributes.

jwiegley commented:
It all looks good to me, though I haven't had the time to try it out yet.

Zimmi48 commented:
> Do you know if there is a way to inherit all attributes from a set? Currently the line `inherit (callPackage ./coq-packages.nix {}) …;` repeats all the attributes.

Not an answer to this question, but we could also go the same route as `ocamlPackages` which recently moved from top-level to `ocaml-ng`, with some aliases kept for compatibility, although I must say that this change confused me until I found where new package sets were hiding.

vbgl commented:
@Zimmi48 Just above (https://github.com/NixOS/nixpkgs/pull/32737#discussion_r157352534) Orivej suggested not to introduce such an attribute (because of confusing naming).

orivej commented:
This PR looks ready to merge, but "coqPackages: wip" suggests that the last commit is work in progress, i.e. unfinished. Do you consider it ready?

> Do you know if there is a way to inherit *all* attributes from a set?

There is `{ one = set; } // another-set`, but it is no used in `all-packages` and there is no equivalent with `inherit`. IMHO explicit inherit of coq attributes is fine.

vbgl committed:
coqPackages: update documentation

vbgl committed:
coqPackages.multinomials: init at 1.0

vbgl commented:
I just cleaned the commits and added some documentation. Also I add the library that is used as an example in the docs.

Zimmi48 commented on doc/languages-frameworks/coq.xml:
```diff
@@ -11,31 +11,53 @@
-    generated using <literal>coq_makefile</literal> so we only have to
+    depends on Coq. It builds on the Mathematical Components library, thus it
+    also takes <literal>mathcomp</literal> as <literal>buildInputs</literal>.
+    Its <literal>makefile</literal> has been generated using
```
Use a capital M here, especially since it is a literal.

Zimmi48 commented on doc/languages-frameworks/coq.xml:
```diff
@@ -11,31 +11,53 @@
+  };
 
   buildInputs = [ coq ];
+  propagatedBuildInputs = [ mathcomp ];
```
Some Coq packages also propagate Coq. Should the rule be not to propagate it? In this case, we should probably update the packages which do not respect this rule.

And how important is it to propagate math-comp? Should Coq package dependencies be always propagated? There are also examples where it is not the case.

jwiegley replied:
I think it should only be propagated if users of that library also (almost always) need to require import from mathcomp. Otherwise, they could specify mathcomp in their own dependencies list, right?

Zimmi48 reviewed (commented):
Kudos for updating the doc!
